### PR TITLE
enhance match attribute filter

### PR DESCRIPTION
**Pull request by nevermore3**
#### What type of PR is this?
enhancement

#### What does this PR do?
close https://github.com/vesoft-inc/nebula/issues/3270

### CODE REVIEW

**Comment by Shylock-Hg** [comment]
Add test cases.

File: tests/tck/features/match/Base.feature [hunk 1]
```diff
@@ -518,6 +518,71 @@ Feature: Basic match
...
+      | [:like "Tony Parker"->"Tim Duncan" @0 {likeness: 95}]        |
+    When executing query:
+      """
+      match (v:player)-[e:like{likeness:"99"}]->()  return e
+      """
+    Then the result should be, in any order, with relax comparison:
+      | e |
+    When executing query:
+      """
+      match (v:player{age:"24"-1})  return v
```
Comment (Shylock-Hg) [inline tests/tck/features/match/Base.feature:581]: This will report error in cypher.

Comment (nevermore3) [reply tests/tck/features/match/Base.feature:578]: done

File: src/parser/MatchSentence.h [hunk 2]
```diff
@@ -163,6 +163,8 @@ class MatchNode final {
 
   const MapExpression* props() const { return props_; }
 
+  MapExpression* props() { return props_; }
+
```
Comment (czpmango) [inline src/parser/MatchSentence.h:167]: Why delete const?

File: src/graph/validator/MatchValidator.cpp [hunk 3]
```diff
@@ -527,8 +525,8 @@ Status MatchValidator::validateUnwind(const UnwindClause *unwindClause,
 }
 
 StatusOr<Expression *> MatchValidator::makeSubFilter(const std::string &alias,
-                                                     const MapExpression *map,
-                                                     const std::string &label) const {
+                                                     MapExpression *map,
+                                                     const std::string &label) {
```
Comment (czpmango) [inline src/graph/validator/MatchValidator.cpp:529]: Expressions related implementation changes should not change the Interface of the Validator.

Comment (nevermore3) [reply src/graph/validator/MatchValidator.cpp:529]: need to rewrite the expression after fold  into the mapexpression

File: tests/tck/features/expression/function/TypeConversion.feature [hunk 4]
```diff
@@ -9,72 +9,81 @@ Feature: TypeConversion Expression
...
+    When executing query:
+      """
+      YIELD [toFloat(true), toFloat(false)] AS yield_toFloat
+      """
+    Then a SemanticError should be raised at runtime: Type error `toFloat(true)'
 
   Scenario: toInteger
     When executing query:
       """
-      YIELD [toInteger(true), toInteger(false), toInteger(1), toInteger(3.14),
```
Comment (Shylock-Hg) [inline tests/tck/features/expression/function/TypeConversion.feature:60]: Why change this case?

Comment (nevermore3) [reply tests/tck/features/expression/function/TypeConversion.feature:60]: toInteger(false), toInteger(true) will report error in opencypher,   other modified cases too

Comment (Shylock-Hg) [reply tests/tck/features/expression/function/TypeConversion.feature:60]: I think you should keep the origin case, and add new one to test the origin function.

Comment (nevermore3) [reply tests/tck/features/expression/function/TypeConversion.feature:60]: i know, but the result of origin case is wrong, and the modified logic test has been covered

**Review by CPWstatic — DISMISSED** [review]
LGTM. Please fix the conflicts.

**Review by yixinglu — APPROVED** [review]
LGTM

**Review by cangfengzhs — APPROVED** [review]
LGTM. But, I have a small question. Is it reasonable to expose `BAT_TYPE` to users? I haven't learned about GQL and opencypher. If they are like this, that's OK. But if they are not like this, is it necessary for us to standardize the `null type` we return to user? After all, our `Value` has many kinds of `null type`.

**Comment by jievince** [comment]
> LGTM. But, I have a small question. Is it reasonable to expose `BAT_TYPE` to users? I haven't learned about GQL and opencypher. If they are like this, that's OK. But if they are not like this, is it necessary for us to standardize the `null type` we return to user? After all, our `Value` has many kinds of `null type`.
Neither of them return `BAD_TYPE`, so the `NULL`, `BAD_TYPE` should be reconsidered

**Comment by czpmango** [comment]
> LGTM. But, I have a small question. Is it reasonable to expose `BAT_TYPE` to users? I haven't learned about GQL and opencypher. If they are like this, that's OK. But if they are not like this, is it necessary for us to standardize the `null type` we return to user? After all, our `Value` has many kinds of `null type`.

All bad nulls is actually just an error with an error message. It may be more reasonable to provide a Value overload with customizable error messages, bcz current expressions’ ability to report and transmit errors is very limited.

**Pull request merged**